ONSdigital/ras-frontstage
# Extract strings

# Motivation and Context
Strings within files that are to be shown to the user can now be wrapped in a `gettext` style function that allows them to be retrieved from a strings file, if an override for the string exists.

This means that UX decisions to alter strings can be made by non-developers, without having to allow them to edit code.

`gettext` itself wasn't used because of it's reliance on the binary compiled format, and extended tooling, which is simple for developers but not for UX team members.  This version is less efficient, but stores all values in a JSON file, and initialises them in memory at bootstrapping of the app.  With the available memory, we can easily accomodate this.

As a bonus, the `gettext` drop in replacement can support string sets for locales, so would make translation of frontstage easier if it becomes necessary in future, which seems like a possibility.

I have also written a small command line tool that will extract the strings from a python file using a simple regex.  This will allow developers to easily create the strings files necessary.

This has been framed in the context of i18n, but is initially intended to simplify alteration of string messages presented to users - locale based options are just a quick-win feature I was able to add.

# What has changed
* Added a class module that can be included in a python file to load strings list for that file
* Added tests for this
* Wrapped strings in the `models.py` file, which defines forms used in the app, for ease of use
* Changed some strings in the first instance, to comply better with GDS
* Added CLI script to find and format wrapped strings in files

# How to test?
1. Start ras-frontstage, with all necessary dependencies also running
2. Ensure forms work
3. Take a string value from frontstage and redefine it in its strings file (`frontstage/i18n/form_messages.json`)
4. Re-start frontstage and confirm that string changes appropriately in UI
5. Usual tests (unit & acceptance) to confirm nothing has broken
6. Optionally, test the CLI util against `models.py`, but bear in mind that it's very much a beta feature - not to be heavily relied upon, only to offer help.

# Links
https://trello.com/c/qCcZrkz5
